### PR TITLE
Fixed #30332 -- Fixed crash of ordering by expressions with params in ArrayAgg and StringAgg.

### PR DESCRIPTION
**Pull request by simone6021**
As per title, also added a couple of scenarios for "plain" ordering in string_agg test.

### CODE REVIEW

**Comment by charettes** [comment]
@simone6021 the CI failures are likely caused by #11171.

File: django/contrib/postgres/aggregates/mixins.py [hunk 1]
```diff
@@ -20,14 +20,22 @@ def resolve_expression(self, *args, **kwargs):
...
+                additional_sql_params.extend(ord_sql_params)
+
+            self.extra['ordering'] = 'ORDER BY ' + ', '.join(ord_clauses)
         else:
             self.extra['ordering'] = ''
-        return super().as_sql(compiler, connection)
+
+        sql, sql_params = super().as_sql(compiler, connection)
+
+        return sql, (sql_params + additional_sql_params)
```
Comment (felixxm) [inline django/contrib/postgres/aggregates/mixins.py]: I think we can move entire `ordering` logic to a separate branch i.e.
```python
if self.ordering:
    ...
    sql, sql_params = super().as_sql(compiler, connection, ordering=(
        'ORDER BY ' + ', '.join(ordering_expr_sql)
    ))
    return sql, sql_params + ordering_params
return super().as_sql(compiler, connection, ordering='')
```

File: django/contrib/postgres/aggregates/mixins.py [hunk 2]
```diff
@@ -20,14 +20,22 @@ def resolve_expression(self, *args, **kwargs):
...
     def as_sql(self, compiler, connection):
+        additional_sql_params = []
+
         if self.ordering:
-            self.extra['ordering'] = 'ORDER BY ' + ', '.join((
-                ordering_element.as_sql(compiler, connection)[0]
-                for ordering_element in self.ordering
-            ))
+            ord_clauses = []
+            for ordering_element in self.ordering:
```
Comment (felixxm) [inline django/contrib/postgres/aggregates/mixins.py]: Few naming suggestions:

`ordering_element` -> `expr`
`ord_sql` -> `expr_sql`
`ord_sql_params` -> `expr_params`
`additional_sql_params` -> `ordering_params`
`ord_clauses` -> `ordering_expr_sql`

File: django/contrib/postgres/aggregates/mixins.py [hunk 3]
```diff
@@ -20,14 +20,22 @@ def resolve_expression(self, *args, **kwargs):
...
+                ord_sql, ord_sql_params = ordering_element.as_sql(compiler, connection)
+                ord_clauses.append(ord_sql)
+                additional_sql_params.extend(ord_sql_params)
+
+            self.extra['ordering'] = 'ORDER BY ' + ', '.join(ord_clauses)
         else:
             self.extra['ordering'] = ''
-        return super().as_sql(compiler, connection)
+
```
Comment (felixxm) [inline django/contrib/postgres/aggregates/mixins.py]: Please chop all unnecessary blank lines.

File: tests/postgres_tests/test_aggregates.py [hunk 4]
```diff
@@ -164,6 +167,10 @@ def test_string_agg_charfield_ordering(self):
             (F('char_field').desc(), 'Foo4;Foo3;Foo2;Foo1'),
             (F('char_field').asc(), 'Foo1;Foo2;Foo3;Foo4'),
             (F('char_field'), 'Foo1;Foo2;Foo3;Foo4'),
+            ('char_field', 'Foo1;Foo2;Foo3;Foo4'),
+            ('-char_field', 'Foo4;Foo3;Foo2;Foo1'),
```
Comment (felixxm) [inline tests/postgres_tests/test_aggregates.py]: These two tests are not related with this fix, please move them to a separate commit.

**Comment by felixxm** [comment]
and please rebase from master.

File: django/contrib/postgres/aggregates/mixins.py (hunk 1 above)
Comment (simone6021) [reply django/contrib/postgres/aggregates/mixins.py]: Done, not a fan of repeated return but code seems more clear now.

File: django/contrib/postgres/aggregates/mixins.py (hunk 2 above)
Comment (simone6021) [reply django/contrib/postgres/aggregates/mixins.py]: Much better names.

File: tests/postgres_tests/test_aggregates.py (hunk 4 above)
Comment (simone6021) [reply tests/postgres_tests/test_aggregates.py]: Done!

File: django/contrib/postgres/aggregates/mixins.py (hunk 3 above)
Comment (simone6021) [reply django/contrib/postgres/aggregates/mixins.py]: Too much consended imho, but probably more in line with django code style.

**Comment by simone6021** [comment]
Thanks @charettes for clarification on CI failures, how to trigger another build in jenkins when issues from #11171 will be solved?

Thanks @felixxm for review and advices, glad to contribute!

**Comment by felixxm** [comment]
@simone6021 Thanks :champagne:
